src/main/java/io/cryostat/core/agent/LocalProbeTemplateService.java

```diff
@@ -0,0 +1,157 @@
+                            TEMPLATE_PATH));
+        }
+        // Sanity check the directory is set up correctly
+        Path probeTemplateDirectory = fs.pathOf(env.getEnv(TEMPLATE_PATH));
+        if (!fs.exists(probeTemplateDirectory)
+                || !fs.isDirectory(probeTemplateDirectory)
+                || !fs.isReadable(probeTemplateDirectory)
+                || !fs.isWritable(probeTemplateDirectory)) {
+            throw new IOException(
+                    String.format(
+                            "Probe template directory %s does not exist, is not a directory, or has incorrect permissions.",
+                            probeTemplateDirectory.toString()));
+        }
+    }
+
+    public void addTemplate(InputStream inputStream, String filename) throws Exception {
+        try (inputStream) {
+            ProbeTemplate template = new ProbeTemplate();
+            // If validation fails this will throw a ProbeValidationException with details
+            template.deserialize(inputStream);
+            Path path = fs.pathOf(env.getEnv(TEMPLATE_PATH), filename);
+            if (fs.exists(path)) {
+                throw new Exception(
+                        String.format(
+                                "Event template \"%s\" already exists", template.getFileName()));
+            }
+            fs.writeString(
+                    path,
+                    new BufferedReader(new InputStreamReader(inputStream, StandardCharsets.UTF_8))
+                            .lines()
+                            .collect(Collectors.joining("\n")));
+        } catch (Exception e) {
+            // rethrow for http handler
+            throw e;
+        }
+    }
+
+    public void deleteTemplate(String templateName) throws Exception {
+        if (!fs.deleteIfExists(fs.pathOf(env.getEnv(TEMPLATE_PATH), templateName))) {
+            throw new IOException(
+                    String.format("Probe template \"%s\" does not exist", templateName));
+        }
+    }
+
+    public String getTemplate(String templateName) throws Exception {
+        Path probeTemplatePath = fs.pathOf(env.getEnv(TEMPLATE_PATH), templateName);
+        ProbeTemplate template = new ProbeTemplate();
+        template.deserialize(fs.newInputStream(probeTemplatePath));
+        return template.serialize();
+    }
+
+    protected List<Path> getLocalTemplates() throws IOException {
+        if (!env.hasEnv(TEMPLATE_PATH)) {
+            return Collections.emptyList();
+        }
+        String dirName = env.getEnv(TEMPLATE_PATH);
+        Path dir = fs.pathOf(dirName);
+        if (!fs.isDirectory(dir) || !fs.isReadable(dir)) {
+            throw new IOException(String.format("%s is not a readable directory", dirName));
+        }
+        try {
+            return fs.listDirectoryChildren(dir).stream()
+                    .map(name -> fs.pathOf(dirName, name))
```

Comment:
`fs.pathOf(dirName, name)` should be replaced by `dir.resolve(name)` since you already have `dir = fs.pathOf(dirName)` in scope. `map(name -> fs.pathOf(dirName, name))` could therefore become `map(dir::resolve)`, I think.